### PR TITLE
fix(aioredis): function is coroutine instead of future when pool is empty in aioredis 1.3

### LINKED ISSUE #3133: AttributeError: 'coroutine' object has no attribute 'add_done_callback'
### Which version of dd-trace-py are you using?

ddtrace 0.57.1

### Which version of pip are you using?

pip 21.3.1

### Which version of the libraries are you using?

aioredis 1.3.1

### What is the result that you get?

```
  File "aioredis/commands/list.py", line 97, in lrange
    return self.execute(b'LRANGE', key, start, stop, encoding=encoding)
  File "ddtrace/contrib/aioredis/patch.py", line 147, in traced_13_execute_command
    task.add_done_callback(_finish_span)
```

### What is the result that you expected?

No errors

### PR DESCRIPTION
**Pull request by mastizada**
Fixes #3133

Problem reproduced: 
[_Example of the error in CI_](https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/13278/workflows/d03dd931-9619-4a9d-8626-ec2fb10aa55c/jobs/948017)

## Commit Message
* aioredis 1.3 execute can be a coroutine

The execute command might be `coroutine` instead of `future` ([source](https://github.com/aio-libs/aioredis-py/blob/v1.3.1/aioredis/pool.py#L191)), that is why it is producing this error:
`AttributeError: 'coroutine' object has no attribute 'add_done_callback'`

This PR creates another future from the coroutine and assigns the callback function to that future.

Build is failing for Python 3.10, but couldn't find out why exactly, might be due to creating the pool in aioredis 1.3 (default behavior of aioredis 2). Might be related to [this](https://github.com/django/channels_redis/issues/278).

## Checklist
- [ ] Added to the correct milestone.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Library documentation is updated.
- [ ] [Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).

PS. Had to do testing in the PR (CI) due to issues with the docker-compose where riot testing (following the documentation) provided `404 page not found` error for all aioredis test cases except `test_patching`.

### CODE REVIEW

File: ddtrace/contrib/aioredis/patch.py
Comment (brettlangdon) [inline ddtrace/contrib/aioredis/patch.py:150]: It might be better to instead do:

```python
if not asyncio.isfuture(task):
    task = asyncio.ensure_future(task)
```

Comment (brettlangdon) [reply ddtrace/contrib/aioredis/patch.py:150]: A comment here on why/when we get a coroutine vs a future would be great

File: tests/contrib/aioredis/test_aioredis.py
Comment (brettlangdon) [inline tests/contrib/aioredis/test_aioredis.py:73]: I know most of the tests here don't have them... but would you be able to add a docstring explaining this test case.

File: tests/contrib/aioredis/test_aioredis.py
Comment (brettlangdon) [inline tests/contrib/aioredis/test_aioredis.py:27]: does this has any particular impact on the other tests?

instead should we introduce a new fixture for a `redis_pool` ?

at the very least we should add a fixture for an `empty_redis_pool` to handle properly setting it up and tearing it down:


```python
@pytest.mark.asyncio
@pytest.fixture
async def empty_redis_pool():
    client = aioredis.create_redis_pool()
    original_pool = getattr(client.connection, "_pool", None)
    if original_pool:
        client.connection._pool = collections.deque(maxlen=1)
    try:
        yield client
    finally:
        client.connection._pool = original_pool
```


What do you think?

File: tests/contrib/aioredis/test_aioredis.py
Comment (brettlangdon) [inline tests/contrib/aioredis/test_aioredis.py:75]: when will this not be defined?

**Review by brettlangdon — COMMENTED** [review]
@mastizada sorry for the long delay on the review, thank you for catching this issue and opening this PR. I left some feedback, let me know what you think.

File: tests/contrib/aioredis/test_aioredis.py
Comment (mastizada) [reply tests/contrib/aioredis/test_aioredis.py:27]: Good idea, that would solve the issue we have at the moment. As for the original fixture, it is creating a pool for aioredis 2+, so, it would make sense to do the same for aioredis 1.3 too. I will add a fixture for empty pool.

File: ddtrace/contrib/aioredis/patch.py
Comment (mastizada) [reply ddtrace/contrib/aioredis/patch.py]: I got this `callable` call idea from other testing cases in this project, `asyncio.isfuture` looks more clear. Changed.
Also added a comment with the reference to the source code where it says when coroutine is returned.

File: tests/contrib/aioredis/test_aioredis.py
Comment (mastizada) [reply tests/contrib/aioredis/test_aioredis.py]: Yes, it was hard to understand what it is testing, added a docstring. Hope it is more clear now. Should I also add info where this is handled?

File: tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json
Comment (brettlangdon) [inline tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]: Is this error going to be consistent?

Is it always going to timeout? Is that part of the underlying issue or a side effect of the specific test case?

Comment (mastizada) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]: So, as you can see in the testing code, we want to have a connection with just 1 pool, then block that pool so that we don't have any available pools left, after that call any method that makes a call to execute. When no pool is left, the execute is a coroutine instead of future and we have confirmed that in commit a62d010 where we get the expected error (before adding ensure_future). However, because we never release the pool, it will wait a long time waiting for a pool to be available (verified in commit 10a8fd6, before adding timeout). After adding timeout, we expect the patch to work successfully and then fail from the timeout while waiting for a free pool.

Ideal way would be to have 2 different tasks where the first one would block the pool, second one will make this requests and only after that the first task will release the block. However, I'm not sure how to manage that case asynchronously in tests, so, timeout was the second best idea. I got the code used for blocking the pool (with (await pool)) and timeout from the aioredis repository's testing where they tested pools for aioredis 1.3.

Comment (mastizada) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]: I don't understand the current error in CI (`failed on teardown with "500 Internal Server Error`), can you help with that?

Comment (mastizada) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]: @brettlangdon any ideas for the error in circleci for aioredis?

Comment (brettlangdon) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]: hey @mastizada so sorry, I took a quick look and it wasn't clear to me. I haven't had the chance to dig in much yet today.

You haven't seen that locally right?

**Comment by brettlangdon** [comment]
@mastizada can you also add a changelog entry?

```
pip install reno
reno new slug-for-the-entry
```

This would be a `fixes:` entry. We usually try to follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) format (when we remember) so maybe `fix(aioredis): <message>` ?

File: tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json
Comment (mastizada) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]: yeah, I can't run testing locally :(
I tried steps written in the Readme when I started working on this, couldn't run tests (without making any changes, just cloned, installed dependencies with docker-compose and tried riot inside docker image using `scripts/ddtest`, it failed).

Comment (brettlangdon) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]: hmm :/ sorry about that, our onboarding/dev docs are really not amazing, definitely some non-obvious internal knowledge there :(

Comment (mastizada) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]: Added changelog too, for the commit message, I can rebase all commits at the end when it will be ready to merge. Also, I'm rebasing with upstream to move these commits in front of commits in master (that is why I use force push).

Comment (brettlangdon) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]: I am trying to run these tests locally and see if I can reproduce the error you are seeing, but my very hastily written riot/ddtrace testing how-to if you were curious.

```
# Run required local services
$ docker-compose up -d testagent redis

# Start our dev vm which has latest `riot` and all the versions of Python we care about
$ scripts/ddtest

# Create base virtual environments (`riot run` can do this for us, but doing explicitly helps speed things up later)
# Note: you only need to do this the first time, or whenever C-extensions need to be rebuilt
$ riot -v generate [-p 3.x -p 3.x ...]

# Run all aioredis tests locally
# Point the tracer to the test agent service via `DD_TRACE_AGENT_URL`
# `--pass-env` is required to get `DD_TRACE_AGENT_URL` into the venvs
# `-s` tells `riot` to skip install `ddtrace` as a dev package (`riot generate` handled this for us)
$ DD_TRACE_AGENT_URL=http://127.0.0.1:9126/ riot -v run -s --pass-env aioredis

# You can scope `riot run` with `-p 3.10` to a specific Python version
$ DD_TRACE_AGENT_URL=http://127.0.0.1:9126/ riot -v run -s -p 3.10 --pass-env aioredis

# You can pass `pytest` args after `--`
$ DD_TRACE_AGENT_URL=http://127.0.0.1:9126/ riot -v run -s --pass-env aioredis -- -k test_closed_connection_pool

# You can use `riot list aioredis` to get a list of venvs, and then pass one of
#   the hashes to `riot shell <hash>` to activate a specific venv if you want
#   to just run `pytest` manually
$ riot list aioredis
$ riot shell <hash>
venv> DD_TRACE_AGENT_URL=http://127.0.0.1:9126/ pytest tests/contrib/aioredis
```

Comment (brettlangdon) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]:
```
[21:57:31] INFO     Running with Interpreter(_hint='3.6')                                                                                          riot.py:653
           INFO     Installing venv dependencies 'pytest-asyncio' 'aioredis' at /root/project/.riot/venv_py3612_pytest-asyncio_aioredis.           riot.py:510
[21:57:37] INFO     Running command 'pytest  tests/contrib/aioredis' in venv '/root/project/.riot/venv_py3612' with environment:                   riot.py:689
                    PYENV_HOOK_PATH=/root/.pyenv/pyenv.d:/usr/local/etc/pyenv.d:/etc/pyenv.d:/usr/lib/pyenv/hooks
                    PYTHONUNBUFFERED=1
                    HOSTNAME=docker-desktop
                    DD_TRACE_AGENT_URL=http://127.0.0.1:9126/
                    PYENV_VERSION=3.9.1:2.7.18:3.5.10:3.6.12:3.7.9:3.8.7:3.10.0
                    PWD=/root/project
                    SIRUN_VERSION=0.1.6
                    HOME=/root
                    LANG=C.UTF-8
                    PYENV_DIR=/root/project
                    TERM=xterm
                    SHLVL=1
                    PYENV_ROOT=/root/.pyenv
                    PATH=/root/project/.riot/venv_py3612_pytest-asyncio_aioredis/bin:/root/project/.riot/venv_py3612_mock_pytest_pytest-mock_cover
                    age_pytest-cov_opentracing_hypothesis/bin:/root/project/.riot/venv_py3612/bin:/root/.pyenv/versions/3.9.1/bin:/root/.pyenv/lib
                    exec:/root/.pyenv/plugins/python-build/bin:/root/.pyenv/shims:/root/.pyenv/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/b
                    in:/sbin:/bin
                    TOX_SKIP_DIST=True
                    NODE_VERSION=node_16.x
                    DD_TESTING_RAISE=1
                    PYTHONPATH=:/root/project:/root/project/.riot/venv_py3612_pytest-asyncio_aioredis/lib/python3.6/site-packages:/root/project/.r
                    iot/venv_py3612_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis/lib/python3.6/site-packages:/root/project/.
                    riot/venv_py3612/lib/python3.6/site-packages.
==================================================================== test session starts =====================================================================
platform linux -- Python 3.6.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /root/project, configfile: tox.ini
plugins: ddtrace-0.58.0.dev24+gaf48a7d19, asyncio-0.16.0, hypothesis-6.31.6, cov-3.0.0, mock-3.6.1
collected 11 items

tests/contrib/aioredis/test_aioredis.py ....FE......                                                                                                   [100%]/root/project/.riot/venv_py3612_mock_pytest_pytest-mock_coverage_pytest-cov_opentracing_hypothesis/lib/python3.6/site-packages/coverage/data.py:129: CoverageWarning: Couldn't use data file '/root/project/.coverage.docker-desktop.280.263612': database disk image is malformed
  data._warn(str(exc))


=========================================================================== ERRORS ===========================================================================
______________________________________________________ ERROR at teardown of test_closed_connection_pool ______________________________________________________
500 Internal Server Error

Server got itself in trouble
========================================================================== FAILURES ==========================================================================
________________________________________________________________ test_closed_connection_pool _________________________________________________________________

single_pool_redis_client = Redis<ConnectionPool<Connection<host=127.0.0.1,port=6379,db=0>>>

>   ???
E   AttributeError: __enter__

/Users/brett.langdon/datadog/dd-trace-py/tests/contrib/aioredis/test_aioredis.py:113: AttributeError
-------------------------------------------------- generated xml file: /root/project/test-results/junit.xml --------------------------------------------------


==================================================================== slowest 10 durations ====================================================================
0.01s teardown tests/contrib/aioredis/test_aioredis.py::test_long_command
0.01s teardown tests/contrib/aioredis/test_aioredis.py::test_basic_request
0.01s teardown tests/contrib/aioredis/test_aioredis.py::test_pipeline_traced_context_manager_transaction
0.01s teardown tests/contrib/aioredis/test_aioredis.py::test_override_service_name
0.01s teardown tests/contrib/aioredis/test_aioredis.py::test_two_traced_pipelines
0.01s teardown tests/contrib/aioredis/test_aioredis.py::test_decoding_non_utf8_args
0.01s teardown tests/contrib/aioredis/test_aioredis.py::test_pipeline_traced
0.01s setup    tests/contrib/aioredis/test_aioredis.py::test_patching
0.01s teardown tests/contrib/aioredis/test_aioredis.py::test_decoding_non_utf8_pipeline_args
0.01s call     tests/contrib/aioredis/test_aioredis.py::test_long_command
================================================================== short test summary info ===================================================================
FAILED tests/contrib/aioredis/test_aioredis.py::test_closed_connection_pool - AttributeError: __enter__
ERROR tests/contrib/aioredis/test_aioredis.py::test_closed_connection_pool
=========================================================== 1 failed, 10 passed, 1 error in 13.59s ===========================================================
Test failed with exit code 1
```


I get this locally as well, this one in particular seemed more clear, no `__enter__`.

Comment (brettlangdon) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]:
```
=========================================================================== ERRORS ===========================================================================
______________________________________________________ ERROR at teardown of test_closed_connection_pool ______________________________________________________
500 Internal Server Error

Server got itself in trouble
========================================================================== FAILURES ==========================================================================
________________________________________________________________ test_closed_connection_pool _________________________________________________________________

single_pool_redis_client = Redis<ConnectionPool<Connection<host=127.0.0.1,port=6379,db=0>>>

    @pytest.mark.asyncio
    @pytest.mark.snapshot
    async def test_closed_connection_pool(single_pool_redis_client):
        """Make sure it doesn't raise error when no free connections are available."""
>       with (await single_pool_redis_client):
E       AttributeError: __enter__

tests/contrib/aioredis/test_aioredis.py:113: AttributeError
```

Comment (brettlangdon) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]: > Added changelog too, for the commit message, I can rebase all commits at the end when it will be ready to merge. Also, I'm rebasing with upstream to move these commits in front of commits in master (that is why I use force push).

We squash merge, so you don't need to worry too much about history on this branch.

The PR title + body are used as the commit title/message, so as long as those following the correct formatting then we are good.

As well, we do require branches to be up to date with `master` before they can be merged, but how it is rebased/merged doesn't matter since we squash merge. (Normally @Mergifyio does it for us, or we use the button in GitHub)

Thanks for asking though!

Comment (brettlangdon) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json:17]: It isn't clear to me the correct path forward with this specific tests. I'll have to think on it a bit.

Basically, we are able to simulate the scenario easily enough? If not, is it worth the hassle of setting up the scenario?


One option we have is, we have implemented a set of tests suites where we run the test suite of the library we have instrumented with our patching enabled in order to make sure we don't cause a test to fail because of something we did.

Example: https://github.com/DataDog/dd-trace-py/blob/af48a7d193f4bff6ff3de022ec67cce1e6bb6d15/.github/workflows/test_frameworks.yml#L53-L93

Maybe this type of test is better / more reliable to grab these types of scenarios.

Comment (mastizada) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json]: Your tutorial for running tests worked. I was able to test locally and fixed most of these issue.
1. It doesn't happen in aioredis 2+ because it just provides an error when pool is not available (another bad change in aioredis for us).
2. Using timeout also cancels the task, which means, the loop (`@pytest.mark.asyncio`) will try to finish it (callback function for ddtrace span results) and get Cancelled Error. Instead, I just created a background task using asyncio.gather after blocking the pool and awaited that after releasing the pool.

To verify it, disabling changes in the patch file gives the expected error that `add_done_callback` is not available for coroutine.

No I have an error without any output (progress? 😄 )

Comment (mastizada) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json]: The error "500 Internal Server Error" is an http error, I'm trying to understand where exactly tests are making an http request. My current assumption, it is from datadog agent at "http://127.0.0.1:9126". Is there a way to check error details for this?

Comment (mastizada) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json]: OMG. Yes, it was datadog testing agent. There was a single comma in the snapshot json file 😨 🔫

Comment (mastizada) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json]: The problem was solved in Python 3.10.1: `bpo-45416` in https://docs.python.org/3/whatsnew/changelog.html
Should we somehow disable check for 3.10 (`select_pys` in riotfile) or add testing for 3.10.1?

Comment (brettlangdon) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json]: > OMG. Yes, it was datadog testing agent. There was a single comma in the snapshot json file 

wow, awesome find! cc @Kyle-Verhoog maybe we need some better error messaging from the test agent?

> updating build version for all tests to 3.10.1)

This is easy enough! https://github.com/DataDog/dd-trace-py/pull/3128

Comment (mastizada) [reply tests/snapshots/tests.contrib.aioredis.test_aioredis.test_closed_connection_pool.json]: Any ideas why it is still trying to build using 3.10.0 in CI?

**Review by brettlangdon — DISMISSED** [review]

**Comment by brettlangdon** [comment]
> I don't know why some other tests cases unrelated to these changes are failing

We do have some flaky tests. I'll manage those (mostly just need to re-run until they pass 😢).

We will need one more review, I'll try to get another team member to look this over today.

**Comment by brettlangdon** [comment]
@Mergifyio backport 0.57

**Review by jd — CHANGES_REQUESTED** [review]

**Review by jd — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**
